pkg/meter/native/provider.go

```
@@ -0,0 +1,124 @@
+	databasev1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/database/v1"
+	"github.com/apache/skywalking-banyandb/banyand/metadata"
+	"github.com/apache/skywalking-banyandb/pkg/logger"
+	"github.com/apache/skywalking-banyandb/pkg/meter"
+)
+
+const (
+	// NativeObservabilityGroupName is the native observability group name.
+	NativeObservabilityGroupName = "_monitoring"
+	defaultTagFamily             = "default"
+	defaultFieldName             = "value"
+	nodeNameTag                  = "node_name"
+)
+
+var log = logger.GetLogger("observability", "metrics", "system")
+
+type provider struct {
+	metadata metadata.Repo
+	scope    meter.Scope
+}
+
+// NewProvider returns a native metrics Provider.
+func NewProvider(scope meter.Scope, metadata metadata.Repo) meter.Provider {
+	return &provider{
+		scope:    scope,
+		metadata: metadata,
+	}
+}
+
+// Counter returns a no-op implementation of the Counter interface.
+func (p *provider) Counter(name string, labelNames ...string) meter.Counter {
+	err := p.createMeasure(name, labelNames...)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failure to createMeasure for Counter %s, labels: %v", name, labelNames)
+	}
+	return noopInstrument{}
+}
+
+// Gauge returns a no-op implementation of the Gauge interface.
+func (p *provider) Gauge(name string, labelNames ...string) meter.Gauge {
+	err := p.createMeasure(name, labelNames...)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failure to createMeasure for Gauge %s, labels: %v", name, labelNames)
+	}
+	return noopInstrument{}
+}
+
+// Histogram returns a no-op implementation of the Histogram interface.
+func (p *provider) Histogram(_ string, _ meter.Buckets, _ ...string) meter.Histogram {
+	return noopInstrument{}
+}
+
+func (p *provider) createMeasure(metric string, labels ...string) error {
+	tags, entityTags := p.getTags(labels)
+	_, err := p.metadata.MeasureRegistry().CreateMeasure(context.Background(), &databasev1.Measure{
```

If the measurement exists, an error "schema.ErrGRPCAlreadyExists" will be returned. You can ignore the error using "errors.Is".